packages/react-ui/src/Uploader.jsx

```diff
@@ -0,0 +1,116 @@
+import { useContext } from 'react'
+import { useUploader, UploaderProvider } from '@w3ui/react-uploader'
+
+const UploaderContext = useContext({})
+
+export const Uploader = ({
+  children,
+  Uploading = DefaultUploading,
+  Errored = DefaultErrored,
+  Done = DefaultDone,
+  ...props
+}) => {
+  const [{ storedDAGShards }, uploader] = useUploader()
+  const [file, setFile] = useState(null)
+  const [dataCid, setDataCid] = useState('')
+  const [status, setStatus] = useState('')
+  const [error, setError] = useState(null)
+
+  const handleUploadSubmit = async e => {
+    e.preventDefault()
+    try {
+      setStatus('uploading')
+      const cid = await uploader.uploadFile(file)
+      setDataCid(cid)
+    } catch (err) {
+      console.error(err)
+      setError(err)
+    } finally {
+      setStatus('done')
+    }
+  }
+  return (
+    <UploaderContext.Provider value={{ storedDAGShards, file, setFile, dataCid, status, error }}>
+      {(status === 'uploading') ? (
+        <Uploading file={file} storedDAGShards={storedDAGShards} />
+      ) : (
+        (status === 'done') ? (
+          error ? <Errored error={error} /> : <Done file={file} dataCid={dataCid} storedDAGShards={storedDAGShards} />
+        ) : (
+          <form onSubmit={handleUploadSubmit} {...props}>
```

Comment:
should this use Uploader.Form?

Reply:
oh, actually this should be inserting `children` here rather than providing any UI - one of the goals of this particular component is to avoid concrete HTML entirely so that consumers have maximum markup freedom - as currently written the `SimpleUploader` example doesn't work at all, lemme update..